2024 Embedded Survey Review (#787)
rust-embedded/wg: surveys/2024-08-survey.md

# Survey questions

This is the first Rust Embedded Working Group community survey. The goal is to gather feedback from users of Rust on Bare Metal or RTOS based embedded devices including microcontrollers.

## Your use of embedded Rust

### How are you currently using embedded Rust?

Type: select all that apply (optional)

- Hobby or personal projects
- School projects
- Work or professional projects
- Not currently using embedded Rust, but have in the past for projects
- I have not used embedded Rust for any projects

### How long have you been using embedded Rust?

Type: select one (optional)

- Zero to one years
- One to two years
- Two to three years
- Three to four years
- Four to five years
- Five to six years
- More than six years

### If you are using embedded Rust for work, how many devices per year is your company shipping using Rust?

Type: select one (optional)

- Unsure or unable to disclose
- 1 to 10 devices per year
- 10 to 100 devices per year
- 100 to 1000 devices per year
- 1000 to 10000 devices per year
- More than 10000 devices per year

> ### What architectures are you using embedded Rust with?

Reviewer: open to the same question as in the suggestion down below, if we should include a "plan to use"  into the question.

Type: select all that apply (optional)

- Arm Cortex-A
- Arm Cortex-M
- Arm Cortex-R
- RISC-V 32-bit
- RISC-V 64-bit
- Xtensa
- Other (open response)

> ### What chip families are you using embedded Rust with?

Reviewer: ```suggestion
### What chip families do you use or plan to use with embedded Rust?
```
That way people interested but not using Rust would potentially not skip the question. Though that would change who we track here, so its debatable if that should be included or not.

Author: Open to it, i'll see if others have feedback for or against before merging!

Reviewer 2: Best to time box it, "or plan to use in the next 12months"

Author: Decided to go with this!

Type: select all that apply (optional)

- ATSAM (Atmel/Microchip)
- ESP32 (Espressif Systems)
- LPC or i.MX RT (NXP)

> - nRF52 (Nordic Semiconductor)

Author: ```suggestion
- nRF (Nordic Semiconductor)
```

Nordic has many more families now, including nRF52, 53, 54, and 91, so let's just go for "nRF"

- RP2xxx (Raspberry Pi)
- STM32 (STMicroelectronics)
- TI (Texas Instruments)
- Other (open response)

### How do you typically use embedded Rust in your projects?

Type: select all that apply (optional)

- Exclusively Rust for firmware
- Majority Rust for firmware, with some C components or libraries
- Majority C for firmware, with some Rust components or libraries
- Exclusively C for firmware

## Typical tools and libraries used

### What frameworks or RTOSs are you using with embedded Rust?

Type: select all that apply (optional)

- Embassy
- ESP-IDF (`std` support on ESP32)
- Hubris OS
- RTIC
- Tock-OS
- Other (open response)

### What tools do you use as part of your development process?

Type: select all that apply (optional)

- Cargo

> - defmt

Author: ```suggestion
- cross
- defmt
```

# Survey questions

This is the first Rust Embedded Working Group community survey. The goal is to gather feedback from users of Rust on Bare Metal or RTOS based embedded devices including microcontrollers.

## Your use of embedded Rust

### How are you currently using embedded Rust?

Type: select all that apply (optional)

- Hobby or personal projects
- School projects
- Work or professional projects
- Not currently using embedded Rust, but have in the past for projects
- I have not used embedded Rust for any projects

### How long have you been using embedded Rust?

Type: select one (optional)

- Zero to one years
- One to two years
- Two to three years
- Three to four years
- Four to five years
- Five to six years
- More than six years

### If you are using embedded Rust for work, how many devices per year is your company shipping using Rust?

Type: select one (optional)

- Unsure or unable to disclose
- Zero devices (only internal or non-production usage)
- 1 to 10 devices per year
- 10 to 100 devices per year
- 100 to 1000 devices per year
- 1000 to 10000 devices per year

> - More than 10000 devices per year

Reviewer 3: Which choice should be selected by a company currently developing, but not yet shipping, a device using rust?

Author: I would probably say the `Zero devices (only internal or non-production usage)` option, the goal here is to gauge "in the field" numbers, IMO!

### What architectures are you using embedded Rust with?

Type: select all that apply (optional)

- Arm Cortex-A
- Arm Cortex-R
- Arm Cortex-M
- RISC-V 32-bit
- RISC-V 64-bit
- Xtensa
- Other (open response)

### What chip families are you using embedded Rust with?

Type: select all that apply (optional)

- ATSAM (Atmel/Microchip)
- ESP32 (Espressif Systems)
- LPC or i.MX RT (NXP)
- nRF (Nordic Semiconductor)
- RP2xxx (Raspberry Pi)
- STM32 (STMicroelectronics)
- TI (Texas Instruments)
- Other (open response)

### How do you typically use embedded Rust in your projects?

Type: select all that apply (optional)

- Exclusively Rust for firmware
- Majority Rust for firmware, with some C components or libraries
- Majority C for firmware, with some Rust components or libraries
- Exclusively C for firmware

## Typical tools and libraries used

> ### What frameworks or RTOSs are you using with embedded Rust?

Reviewer 2: Worth an explicit mention of Zephyr?

Author: Hmm, that's a good point, maybe mention "other C or C++ based RTOSs, such as Zephyr, FreeRTOS, mbed", though that might be possible for people to just write in for "Other"?

Type: select all that apply (optional)

- Embassy
- ESP-IDF (`std` support on ESP32)
- Hubris OS
- RTIC
- Tock-OS
- Other (open response)

### What tools do you use as part of your development process?

Type: select all that apply (optional)

- Cargo
- cross
- defmt

> - probe-rs

Reviewer 2: espflash?

Author: Added in 1449588

# Survey questions

This is the first Rust Embedded Working Group community survey. The goal is to gather feedback from users of Rust on Bare Metal or RTOS based embedded devices including microcontrollers.

## Your use of embedded Rust

### How are you currently using embedded Rust?

Type: select all that apply (optional)

- Hobby or personal projects

> - School projects

Reviewer 4: School or university?

Author: Updated to "School or University"

- I have not used embedded Rust for any projects

### How long have you been using embedded Rust?

Type: select one (optional)

- Zero to one years
- One to two years
- Two to three years
- Three to four years
- Four to five years
- Five to six years
- More than six years

### If you are using embedded Rust for work, how many devices per year is your company shipping using Rust?

Type: select one (optional)

- Unsure or unable to disclose
- Zero devices (only internal or non-production usage)
- 1 to 10 devices per year
- 10 to 100 devices per year
- 100 to 1000 devices per year
- 1000 to 10000 devices per year
- More than 10000 devices per year

### What architectures are you using embedded Rust with?

Type: select all that apply (optional)

- Arm Cortex-A
- Arm Cortex-R
- Arm Cortex-M
- RISC-V 32-bit
- RISC-V 64-bit
- Xtensa
- Other (open response)

### What chip families are you using embedded Rust with?

Type: select all that apply (optional)

- ATSAM (Atmel/Microchip)
- ESP32 (Espressif Systems)
- LPC or i.MX RT (NXP)
- nRF (Nordic Semiconductor)
- RP2xxx (Raspberry Pi)
- STM32 (STMicroelectronics)
- TI (Texas Instruments)

> - Other (open response)

Reviewer 4: Any chance of adding an [Ambiq](https://ambiq.com/) option? It's one of the best supported SoCs in Tock so it would be interesting to list

Author: Added! I hadn't seen it used much personally, but if it is popular in Tock happy to add it!

### How do you typically use embedded Rust in your projects?

Type: select all that apply (optional)

- Exclusively Rust for firmware
- Majority Rust for firmware, with some C components or libraries
- Majority C for firmware, with some Rust components or libraries

> - Exclusively C for firmware

Reviewer 4: What about a C++ option as well? I know it's similar, but it would be interesting to compare the two

Author: I will rewrite these to not be specific to C, this was an oversight!

## Typical tools and libraries used

### What frameworks or RTOSs are you using with embedded Rust?

Type: select all that apply (optional)

- Embassy
- ESP-IDF (`std` support on ESP32)
- Hubris OS
- RTIC
- Tock-OS
- Other (open response)

### What tools do you use as part of your development process?

Type: select all that apply (optional)

- Cargo
- cross
- defmt
- espflash
- probe-rs
- Rust Analyzer

> - Other (open response)

Reviewer 4: What about including [bindgen](https://github.com/rust-lang/rust-bindgen) here?

It's really useful if interacting with existing C code

Author: Added `bindgen` and `cbindgen`

- Unsure or unable to disclose
- Zero devices (only internal or non-production usage)
- 1 to 10 devices per year
- 10 to 100 devices per year
- 100 to 1000 devices per year
- 1000 to 10000 devices per year
- More than 10000 devices per year

### What architectures are you planning to use embedded Rust with in the next 12 months?

Type: select all that apply (optional)

- Arm Cortex-A
- Arm Cortex-R
- Arm Cortex-M
- RISC-V 32-bit
- RISC-V 64-bit
- Xtensa
- Other (open response)

### What chip families are you planning to use embedded Rust with in the next 12 months?

Type: select all that apply (optional)

- Apollo (Ambiq)
- ATSAM (Atmel/Microchip)
- ESP32 (Espressif Systems)
- LPC or i.MX RT (NXP)
- nRF (Nordic Semiconductor)
- RP2xxx (Raspberry Pi)
- STM32 (STMicroelectronics)
- TI (Texas Instruments)
- Other (open response)

### How do you typically use embedded Rust in your firmware projects?

Type: select all that apply (optional)

- Exclusively Rust for firmware
- Majority Rust for firmware, with some components or libraries in other languages (C/C++/Ada/Zig)
- Majority other languages (C/C++/Ada/Zig) for firmware, with some Rust components or libraries

> - Exclusively other languages (C/C++/Ada/Zig) for firmware

Reviewer 4: Is it worth adding a question after this to get an idea of what languages (C/C++/Ada/Zig) people are using with Rust? It might help guide future projects

Author: I'm a little hesitant to add more questions, but I'll discuss it tomorrow in the next meeting where we wrap up the review for this!

Author: Added!

## Typical tools and libraries used

### What frameworks or RTOSs are you using with embedded Rust?

Type: select all that apply (optional)

- Embassy
- ESP-IDF (`std` support on ESP32)
- Hubris OS
- RTIC
- Tock-OS
- Non-Rust based RTOSs (such as Zephyr, FreeRTOS, RIOT-OS, etc.)
- Other (open response)

### What tools do you use as part of your development process?

Type: select all that apply (optional)

- bindgen
- Cargo
- cbindgen
- cross
- defmt
- espflash
- probe-rs
- Rust Analyzer
- Other (open response)

> ### What are the most common crates or libraries you use in your embedded projects?

Reviewer 5: Do you perhaps want to enumerate a few popular choices (e.g. postcard :) ) here? If there is a predetermined list, it usually gets more answers than if there is just an open text field, people are lazy to type, unsure how to separate multiple responses etc. Not to mention that it's much easier to analyze.

Author: Decided against! We will reconsider in future surveys

Type: free text (optional)

## Personal feedback

> ### What are your biggest complaints or blockers for using Rust in your embedded projects?

Reviewer 5: Same here.

Author: Decided against! We will reconsider in future surveys

# Survey questions

> This is the first Rust Embedded Working Group community survey. The goal is to gather feedback from users of Rust on Bare Metal or RTOS based embedded devices including microcontrollers.

Reviewer 6: ```suggestion
This is the first Rust Embedded Working Group community survey. The goal is to gather feedback from users of Rust on bare-metal or RTOS based embedded devices including microcontrollers.
```

## Your use of embedded Rust

### Where do you plan to use embedded Rust in the next 12 months?

Type: select all that apply (optional)

- Hobby or personal projects

> - School or University projects

Reviewer 6: ```suggestion
- School or university projects
```

- Work or professional projects
- Not currently using embedded Rust, but have in the past for projects
- I have not used embedded Rust for any projects

### How long have you been using embedded Rust?

Type: select one (optional)

- Zero to one years
- One to two years
- Two to three years
- Three to four years
- Four to five years
- Five to six years
- More than six years

### If you are using embedded Rust for work, how many embedded devices per year is your company shipping using Rust?

Type: select one (optional)

- Unsure or unable to disclose
- Zero devices (only internal or non-production usage)
- 1 to 10 devices per year
- 10 to 100 devices per year

> - 100 to 1000 devices per year
> - 1000 to 10000 devices per year
> - More than 10000 devices per year

Reviewer 6: ```suggestion
- 100 to 1,000 devices per year
- 1,000 to 10,000 devices per year
- More than 10,000 devices per year
```
